node/core/parachains-inherent/src/lib.rs

```diff
@@ -29,32 +29,33 @@ use polkadot_node_subsystem::{
 	errors::SubsystemError, messages::ProvisionerMessage, overseer::Handle,
 };
 use polkadot_primitives::v2::{Block, Hash, InherentData as ParachainsInherentData};
-use sp_blockchain::HeaderBackend;
 use sp_runtime::generic::BlockId;
-use std::time;
+use std::{sync::Arc, time};
 
 pub(crate) const LOG_TARGET: &str = "parachain::parachains-inherent";
 
 /// How long to wait for the provisioner, before giving up.
 const PROVISIONER_TIMEOUT: time::Duration = core::time::Duration::from_millis(2500);
 
 /// Provides the parachains inherent data.
-pub struct ParachainsInherentDataProvider {
-	inherent_data: ParachainsInherentData,
+pub struct ParachainsInherentDataProvider<C: sp_blockchain::HeaderBackend<Block>> {
+	pub client: Arc<C>,
+	pub overseer: polkadot_overseer::Handle,
+	pub parent: Hash,
 }
 
-impl ParachainsInherentDataProvider {
-	/// Create a [`Self`] directly from some [`ParachainsInherentData`].
-	pub fn from_data(inherent_data: ParachainsInherentData) -> Self {
-		Self { inherent_data }
+impl<C: sp_blockchain::HeaderBackend<Block>> ParachainsInherentDataProvider<C> {
+	/// Create a new [`Self`].
+	pub fn new(client: Arc<C>, overseer: polkadot_overseer::Handle, parent: Hash) -> Self {
+		ParachainsInherentDataProvider { client, overseer, parent }
 	}
 
 	/// Create a new instance of the [`ParachainsInherentDataProvider`].
-	pub async fn create<C: HeaderBackend<Block>>(
-		client: &C,
+	pub async fn create(
+		client: Arc<C>,
 		mut overseer: Handle,
 		parent: Hash,
```

Comment:
```suggestion
	async fn create(&self
```

Reply:
Revert as this would need `self`to be `mut` which in the end would need `self` in `InherentDataProvider` to be mut which feels wrong to change for this one case.